src/esphome/sensor/cse7766.cpp

```diff
@@ -101,56 +95,79 @@ void CSE7766Component::parse_data_() {
+    // ESP_LOGV(TAG, "CSE7766 reports abnormal hardware: (0x%02X)", byte);
+    if ((header1 >> 3) & 1) {
+      ESP_LOGV(TAG, "  Voltage cycle exceeds range.");
+      voltage_ok = false;
+    }
+    if ((header1 >> 2) & 1) {
+      ESP_LOGV(TAG, "  Current cycle exceeds range.");
+      current_ok = false;
+    }
+    if ((header1 >> 1) & 1) {
+      ESP_LOGV(TAG, "  Power cycle exceeds range.");
+      power_ok = false;
+    }
+    if ((header1 >> 0) & 1) {
+      ESP_LOGV(TAG, "  Coefficient storage area is abnormal.");
+      return;
+    }
+  }
 
-  if ((adj >> 6 != 0) && voltage_cycle != 0 &&
-      // voltage cycle exceeds range
-      ((header1 >> 3) & 1) == 0) {
+  if ((adj & 0x40) == 0x40 && voltage_ok && current_ok) {
     // voltage cycle of serial port outputted is a complete cycle;
-    float voltage = voltage_calib / float(voltage_cycle);
-    this->voltage_acc_ += voltage * d;
+    this->voltage_acc_ += voltage_calib / float(voltage_cycle);
+    this->voltage_counts_ += 1;
   }
-
-  if ((adj >> 5 != 0) && power_cycle != 0 && current_cycle != 0 &&
-      // current cycle exceeds range
-      ((header1 >> 2) & 1) == 0) {
-    // indicates current cycle of serial port outputted is a complete cycle;
-    float current = current_calib / float(current_cycle);
-    this->current_acc_ += current * d;
+   
+  float power = 0;
+  if ((adj & 0x10) == 0x10 && power_ok) {
+    // power cycle of serial port outputted is a complete cycle;
+    power = power_calib / float(power_cycle);
+    this->power_acc_ += power;
+    this->power_counts_ += 1;
   }
 
-  if ((adj >> 4 != 0) && power_cycle != 0 &&
-      // power cycle exceeds range
-      ((header1 >> 1) & 1) == 0) {
-    // power cycle of serial port outputted is a complete cycle;
-    float active_power = power_calib / float(power_cycle);
-    this->power_acc_ += active_power * d;
+  if ((adj & 0x20) == 0x20 && current_ok && voltage_ok && power != 0.0) {
+    // indicates current cycle of serial port outputted is a complete cycle;
+    this->current_acc_ += current_calib / float(current_cycle);
+    this->current_counts_ += 1;
   }
 }
 void CSE7766Component::update() {
-  const uint32_t now = millis();
-  uint32_t d = now - this->last_update_;
-  this->last_update_ = now;
-  float voltage = this->voltage_acc_ / d;
-  float current = this->current_acc_ / d;
-  float power = this->power_acc_ / d;
+  
+  float voltage = this->voltage_counts_ > 0 ? this->voltage_acc_ / this->voltage_counts_ : 0.0;
+  float current = this->current_counts_ > 0 ? this->current_acc_ / this->current_counts_ : 0.0;
+  float power = this->power_counts_ > 0 ? this->power_acc_ / this->power_counts_ : 0.0;
+
+  ESP_LOGD(TAG, "Got voltage_acc=%.2f current_acc=%.2f power_acc=%.2f", this->voltage_acc_, this->current_acc_, 
+           this->power_acc_);
+  ESP_LOGD(TAG, "Got voltage_counts=%d current_counts=%d power_counts=%d", this->voltage_counts_, this->current_counts_, 
+           this->power_counts_);
```

Comment:
The user doesn't care about the acc and count numbers, please make these verbose or remove them.

Reply:
I have to agree with you for the most part on your (new) version of the code.  In fact, the only problem I had was it was reading current (very small) when it should have been 0.

The spec sheet (see Figure 5, 6 and 7) says that bit3 and bit2 must be 0 for all readings.  

For power, bit1 must be 0, otherwise power is 0.
For current, if measured power is 0 then current is 0.

I saw that some other people are using the cse7759B.  The only datasheet I found for it was in Chinese, but the figures 5,6 and 7 look to be the same.  I didn't really have the time for a detailed analysis, but I will look later tonight and fix the items above.

Thanks for being patient with me here as this GitHub stuff is new to me.  Getting someone else's coding style right is another story.  Python though - I cant help you with!

Fix is on the way